org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java

```diff
@@ -1736,106 +1745,61 @@ public Optional<MetaIndex> getMetaIndex() {
 	}
 }
 
-void scheduleForMetaIndexUpdate(Index index) {
-	synchronized(this.metaIndexUpdates){
-		if (this.metaIndexUpdates.contains(index)) {
-			if (VERBOSE) {
-				Util.verbose("-> already waiting for meta-index update for " + index); //$NON-NLS-1$
-			}
-			return;
-		}
-		this.metaIndexUpdates.add(index);
-	}
-	requestIfNotWaiting(new MetaIndexUpdateRequest());
```

Comment:
Avoiding to put a metaindex-related request into `awaitingJobs` from within `indexerLoop()` certainly avoids the bug as I observed it.